### PR TITLE
Set timeout for http server

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

To make http server resilient

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
There are security issues like a malicious user can attack the server by using up all the connections if no timeout set.

**Comment by hzxuzhonghu** [comment]
@ramaraochavali If we declare a const for timeout, the constants package maybe imported everywhere. I did not see any project do it this way.

**Comment by hzxuzhonghu** [comment]
FYI: https://blog.cloudflare.com/exposing-go-on-the-internet/

**Review by ramaraochavali — APPROVED** [review]
I did not mean constants in a

**Comment by hzxuzhonghu** [comment]
@istio/wg-security-maintainers ptal

**Review by shankgan — APPROVED** [review]

**Comment by shankgan** [comment]
@hzxuzhonghu - did you try using ReadHeaderTimeout? Also, is the 30 sec timeout large enough for XDS?

**Comment by hzxuzhonghu** [comment]
xds is through grpc, this is for the http server.  `ReadHeaderTimeout` and `ReadTimeout` have slight difference, both works here,  `ReadTimeout` is the maximum duration for reading the entire request, including the body. I think it is safer, we have no large body sent to server.

**Comment by howardjohn** [comment]
We do actually serve grpc using the http2 stack when using the port
multiplex (ie set gRPC port to "" and it will serve on the http server).
should make sure timeout doesn't apply to that (or applies to first byte vs
full stream)

**Comment by hzxuzhonghu** [comment]
Let me look into that

File: security/pkg/stsservice/server/server.go [hunk 1]
```diff
@@ -81,8 +82,10 @@ func NewServer(config Config, tokenManager security.TokenManager) (*Server, erro
 	mux.HandleFunc(TokenPath, s.ServeStsRequests)
 	mux.HandleFunc(StsStatusPath, s.DumpStsStatus)
 	s.stsServer = &http.Server{
-		Addr:    fmt.Sprintf("%s:%d", config.LocalHostAddr, config.LocalPort),
-		Handler: mux,
+		Addr:        fmt.Sprintf("%s:%d", config.LocalHostAddr, config.LocalPort),
+		Handler:     mux,
+		IdleTimeout: 90 * time.Second, // matches http.DefaultTransport keep-alive timeout
```
Comment (ryan-dayrit) [inline security/pkg/stsservice/server/server.go:87]: Does it make sense for the 90/30 to come from config.IdleTimeout/config.ReadTimeout vs having it baked into the compiled code?

Comment (hzxuzhonghu) [reply security/pkg/stsservice/server/server.go:87]: @ryan-dayrit What do you mean?

Comment (ryan-dayrit) [reply security/pkg/stsservice/server/server.go:87]: - Extending the config struct to have these IdleTimeout / ReadTimeout props. 
- Setting default vals to 90 / 30 respectively 

```
IdleTimeout: config.IdleTimeoutSecs * time.Second
ReadTimeout: config.ReadTimeoutSecs * time.Second
```

**Comment by hzxuzhonghu** [comment]
@howardjohn From https://golang.org/src/net/http/server.go#L937  the readtimeout and readheaderTimeout is same if only one is specified.  And it does not influence http2 stream, as we have enabled keep alive by default 30s.

**Pull request merged**